### ISSUE TITLE
Strange warped image results

### ISSUE THREAD

**Issue by riccardosamperna** [voxelmorph/voxelmorph#82]
Dear Adrian, 

first of all thanks a lot for your work. I really appreciate your decision of making it public and your availability in answering questions.

I am currently trying to use the code for the registration of pre- and post-contrast breast mri volumes. I am using the miccai2018 network and, as I read in other posts, I normalized the intensity of the images in the range [0, 1]. I do not use any affine alignment because in principle pre- and post-contrast images should be already affinely aligned. The images are non-isotropic with a spacing of approximately [0.6, 0.6, 1.3].

I lowered the learning rate because all the losses were going to NaN and I tried to run the miccai network both with the original parameter for image_sigma and prior_lambda and also with slightly different ones (0.07 and 1 respectively) but all the losses stay constant and I get weird warped images.

Do you know what it can be possibly go wrong? Am I missing something?

I hope I gave enough information about the problem and please feel free to ask any clarification.

**Comment by adalca** [comment]
Hi @riccardosamperna ,

I'm not sure what pre and post-contrast images look like, but from the sound of it i'm not sure the MSE is appropriate for the image matching term (as the miccai model uses). It might be more appropriate to use CC (which is available in the losses file) or MI (which I could send you) depending on the images.

Can you post a screenshot of your images?

**Comment by riccardosamperna** [comment]
Below you can find an example, the idea is to register the post-contrast image to the pre-contrast one.
Pre-contrast
![pre-contrast](https://user-images.githubusercontent.com/6599726/62552337-c3f3ed80-b86d-11e9-9d3b-92496c9def3e.png)
Post-contrast
![post_contrast](https://user-images.githubusercontent.com/6599726/62552344-c6564780-b86d-11e9-8c21-8d6551980899.png)
Warped post-contrast
![warped](https://user-images.githubusercontent.com/6599726/62552351-c9513800-b86d-11e9-950f-489c84f67b94.png)
Deformation-field
![deformation](https://user-images.githubusercontent.com/6599726/62552359-cbb39200-b86d-11e9-8482-595939866a7f.png)

**Comment by adalca** [comment]
While ultimately MI would likely be most robust, these images look like they might work with CC.
I would encourage you to try the CVPR voxelmorph with CC (and the CC hyper-parameter value) simply because we've tested CC for the CVPR model quite a bit and it seems to work well. If that works, it should be quite easy to switch to the MICCAI model.

**Comment by riccardosamperna** [comment]
I have already tried to switch to an NCC loss with the MICCAI network but I am having similar results as before. I will try the code from CVPR and I will keep you posted about any improvement. Thanks a lot for your reply.

**Comment by adalca** [comment]
ok, let me know -- remember that by trying NCC you should dramatically change the hyper-parameter. 
For our brain data, in the CVPR version, for example, if you use MSE you should use a regularization parameter of 0.01, but with CC then 1.0 is a good start. So you'd need to make some changes in the MICCAI version as well.

**Comment by riccardosamperna** [comment]
When I tried the MICCAI version with NCC, I indeed made some changes to the hyperparameters (I used (1.0, 1.0) as loss weights) based on what I saw in the CVPR one but I was still getting large deformations. Currently I am testing the CVPR one with reg_param of 1.5 and it seems to be doing a much better job.

I have a question about the losses, I attached a txt file where you can see the values of the losses in ~150 epochs. It looks like that all the losses, both for training and validation, are not changing or have just small variations, is this behaviour normal? I have seen it in all runs I made so far, similar behaviour when I was using MSE loss.

[07082019T162303_log.txt](https://github.com/voxelmorph/voxelmorph/files/3480346/07082019T162303_log.txt)

**Comment by adalca** [comment]
- if the CVPR version works and the MICCAI version does not with NCC, I think it just requires a better choice of hyper-parameters. Perhaps we'll work at providing specific guidance for MICCAI version with NCC (my first guess wouldn't beed to take lambda from 10 to 1000 while keeping image sigma the same, I think around 0.02).

- Because the batch size is usually only 1, we set the learning rate to be fairly small, and the losses will change seemingly little and very noisly. I would recommend avoiding looking at the losses numerically, it really tends to be misleading because we tend to try to make conclusions from any noisy pattern we see -- I'd recommend plotting them and trying to estimate the overall trend.

**Comment by hfahad** [comment]
Dear Adrian,
I need to ask one thing Can i use Grad('l2') loss with MSE in MICCAI version ?
Because in MICCAI paper you use KL loss with MSE

**Comment by adalca** [comment]
@hfahad,

The MICCAI model gives you diffeomorphisms and probabilities -- which of these do you need, if any? If you want both, the KL loss is a clean thing to use, and this includes a smoothing factor since the KL term includes a graph laplacian. 

However, if all you desire are diffeomorphisms, then you could just avoid using the sampling layer that adds probabilities, and use Grad('l2') on the final displacement after diffeomorphisms. 

If you'd like to continue this discussion, though, please start a new github issue, to not distract from @riccardosamperna's questions.

**Comment by riccardosamperna** [comment]
> Because the batch size is usually only 1, we set the learning rate to be fairly small, and the losses will change seemingly little and very noisly. I would recommend avoiding looking at the losses numerically, it really tends to be misleading because we tend to try to make conclusions from any noisy pattern we see -- I'd recommend plotting them and trying to estimate the overall trend.

I followed your advice (plotted the csv) and I noticed that indeed the loss and the spatial transformer loss are going down, only the flow_loss has an up trend (?). Have you used any kind of learning rate decay strategy during training?

**Comment by adalca** [comment]
@riccardosamperna it's fine for some part of the loss to go up as long as the overall loss goes down.
I've tried learning rate decay before and it can help, but honestly having lr of 1e-4 does most of the job without any more fiddling usually needed.

**Comment by riccardosamperna** [comment]
I am already getting much better results compared to what I posted in the first place. A strong regularization of the deformation field helped in the CVPR network, soon I will try to see also what happens with the MICCAI version. Thanks for your replies and your support @adalca. For the moment I will close the issue, I will keep on posting here if I have any more questions.

**Comment by adalca** [comment]
@riccardosamperna glad to hear it!

**Comment by riccardosamperna** [comment]
Dear @adalca,
I am back with a somewhat unrelated question. The CVPR network I trained for my problem seems to be working now but I would like to investigate the problem of the negative Jacobians I read about in other posts and in your MICCAI paper. Have you checked the influence of this effect on the CVPR network? Can you share any code you used to actually calculate this effect?

**Comment by adalca** [comment]
Hi @riccardosamperna 

Sorry, I'm a bit confused -- are you asking if we counted folding voxels for the CVPR network? Yes, we did (see TMI paper tables), and the count is similar to the calssical methods like ANTs. But you can add integration layers at the end of the CVPR model (and tine-tune the training) and it will dramatically limit these folding voxels. Is your goal to limit these?

**Comment by riccardosamperna** [comment]
Thanks for your reply. Apologies for the confusion, I went back to the MICCAI paper and I checked more carefully the tables (Voxelmorph (CC) refers to the CVPR paper). 
Indeed I would like to limit this folding effect as much as possible. At the moment I have only a visual clue of what is happening but, first of all, I wanted to verify and quantify this folding effect numerically (before adding integration layers at the end of the CVPR model, as you suggested).  How did you calculate the number of folding voxels or equivalently the number of |J| <= 0 from Table 1?

**Comment by adalca** [comment]
Sure, here's some code snippet to compute the determinant, you would need to count the negative entries. Note that computing the jacobian is not well defined on a discrete grill, and there are various ways to do it (e.g. based on various ways to approximate the gradient). There is also code in other libraries like `SimpleITK`.

```
def flow_jacdet(flow):

    vol_size = flow.shape[:-1]
    grid = np.stack(nd.volsize2ndgrid(vol_size), len(vol_size))  #volshape2grid_3d(vol_size)
    J = np.gradient(flow + grid)

    dx = J[0]
    dy = J[1]
    dz = J[2]

    Jdet0 = dx[:,:,:,0] * (dy[:,:,:,1] * dz[:,:,:,2] - dy[:,:,:,2] * dz[:,:,:,1])
    Jdet1 = dx[:,:,:,1] * (dy[:,:,:,0] * dz[:,:,:,2] - dy[:,:,:,2] * dz[:,:,:,0])
    Jdet2 = dx[:,:,:,2] * (dy[:,:,:,0] * dz[:,:,:,1] - dy[:,:,:,1] * dz[:,:,:,0])

    Jdet = Jdet0 - Jdet1 + Jdet2

    return Jdet
```

**Comment by riccardosamperna** [comment]
Great, thanks a lot, this is really helpful.

I thought about making a numerical approximation of the Jacobian but I was not sure about which method to use to have a reliable approximation. A similar approximation is also used for the calculation of the "diffusion regularizer on the spatial gradients of displacement" (for the gradient in this case, in the Grad loss), is it correct? Have you tried to add a "naive" regularizer loss term (a condition on the Jacobian determinant) for the folding effect in the CVPR network? If so, how does it perform compared to the MICCAI version?

**Comment by adalca** [comment]
We haven't added a Jacobian loss term, although other researchers have in literature. We found this to be unnecessary with networks -- going through a SVF (as in the MICCAI version, but you could also just modify the CVPR model easy by adding an integration layer) essentially guarantees very few folding voxels, and this seems clearer than imposing another loss that would need to be tuned, etc, especially given that integration layers are very cheap compared to the rest of the network (as we analyze in the MedIA paper)

**Comment by riccardosamperna** [comment]
Thank you for patiently answering all my questions @adalca, I really appreciate your availability and the clear explanations.

**Comment by adalca** [comment]
glad to help!

**Comment by Tara-Liu** [comment]
> 当然，这里有一些代码片段来计算行列式，你需要计算负项。请注意，计算雅可比矩阵在离散网格上定义不明确，并且有多种方法可以做到这一点（例如，基于各种方法来近似梯度）。其他库中也有代码，例如`SimpleITK`。
> 
> ```
> def flow_jacdet(flow):
> 
>     vol_size = flow.shape[:-1]
>     grid = np.stack(pynd.ndutils.volsize2ndgrid(vol_size), len(vol_size))  
>     J = np.gradient(flow + grid)
> 
>     dx = J[0]
>     dy = J[1]
>     dz = J[2]
> 
>     Jdet0 = dx[:,:,:,0] * (dy[:,:,:,1] * dz[:,:,:,2] - dy[:,:,:,2] * dz[:,:,:,1])
>     Jdet1 = dx[:,:,:,1] * (dy[:,:,:,0] * dz[:,:,:,2] - dy[:,:,:,2] * dz[:,:,:,0])
>     Jdet2 = dx[:,:,:,2] * (dy[:,:,:,0] * dz[:,:,:,1] - dy[:,:,:,1] * dz[:,:,:,0])
> 
>     Jdet = Jdet0 - Jdet1 + Jdet2
> 
>     return Jdet
> ```

What are the parameters of this function? Is the return value of warped_seg = transform_model.predict([moving_seg, warp]).squeeze() warped_seg?

**Comment by adalca** [comment]
This computes the jacobian of a warp field, so the input to flow_jacdet is the warp field itself

**Comment by CharlieCHHQ** [comment]
Dear Adrian, 

I have been encountering the same issues as @riccardosamperna, so I tried using NCC with the earlier version CVPR. I was able to get a much better image, but with little change between my moved and moving. The reason I think MSE doesn't work well with with my dataset, is because of the intensity range difference between the moving and fixed images. Regardless of intensity scaling between 0 and 1, the majority of values for the moving and fixed are different (for example, moving majority values: 0 to 0.1, fixed majority values: 0 to 0.6), which might be the reason for generating strange warped images. I have been testing various preprocessing methods, removing outliers, histogram matching, switching to NCC. However, none of them have proven to be very effective. May I please try the Mutual Information (MI) you mentioned before? Thank you for making VoxelMorph available!

Sincerely,
Charlie
